src/query/storages/system/src/tables_table.rs

```diff
@@ -245,39 +245,44 @@ where TablesTable<T, U>: HistoryAware
-                find_eq_filter(&expr, &mut |col_name, scalar| {
-                    if col_name == "database" {
-                        if let Scalar::String(database) = scalar {
-                            if !db_name.contains(database) {
-                                db_name.push(database.clone());
+                invalid_optimize = find_eq_or_filter(
+                    &expr,
+                    &mut |col_name, scalar| {
+                        if col_name == "database" {
+                            if let Scalar::String(database) = scalar {
+                                if !db_name.contains(database) {
+                                    db_name.push(database.clone());
+                                }
                             }
-                        }
-                    } else if col_name == "table_id" {
-                        match check_number::<_, u64>(
-                            None,
-                            &FunctionContext::default(),
-                            &Expr::<usize>::Constant {
-                                span: None,
-                                scalar: scalar.clone(),
-                                data_type: scalar.as_ref().infer_data_type(),
-                            },
-                            &BUILTIN_FUNCTIONS,
-                        ) {
-                            Ok(id) => tables_ids.push(id),
-                            Err(_) => invalid_tables_ids = true,
-                        }
-                    } else if col_name == "name" {
-                        if let Scalar::String(t_name) = scalar {
-                            if !tables_names.contains(t_name) {
-                                tables_names.push(t_name.clone());
+                        } else if col_name == "table_id" {
+                            match check_number::<_, u64>(
+                                None,
+                                &FunctionContext::default(),
+                                &Expr::<usize>::Constant {
+                                    span: None,
+                                    scalar: scalar.clone(),
+                                    data_type: scalar.as_ref().infer_data_type(),
+                                },
+                                &BUILTIN_FUNCTIONS,
+                            ) {
+                                Ok(id) => tables_ids.push(id),
+                                Err(_) => invalid_tables_ids = true,
+                            }
+                        } else if col_name == "name" {
+                            if let Scalar::String(t_name) = scalar {
+                                if !tables_names.contains(t_name) {
+                                    tables_names.push(t_name.clone());
+                                }
                             }
                         }
-                    }
-                    Ok(())
-                });
+                        Ok(())
```

Comment:
incorrect result.

eg.

```
table = 'a' or database ='b' or num_rows >= 0
```

Reply:
This will hits this code:

```rust
if lfunc.signature.name == "eq" && rfunc.signature.name == "eq" {
match (largs.as_slice(), rargs.as_slice()) {
  (
      [Expr::ColumnRef { id, .. }, Expr::Constant { .. }]
      | [Expr::Constant { .. }, Expr::ColumnRef { id, .. }],
      [Expr::ColumnRef { id: rid, .. }, Expr::Constant { .. }]
      | [Expr::Constant { .. }, Expr::ColumnRef { id: rid, .. }],
  ) => id != rid,
  _ => false,
}
} else {
false
}
```

I test on local:

```
root@localhost:8000/default>  select name,database,num_rows from system.tables where  database = 'default' or name='t' or num_rows>=0;

┌──────────────────────────────────────┐
│  name  │ database │     num_rows     │
│ String │  String  │ Nullable(UInt64) │
├────────┼──────────┼──────────────────┤
│ t      │ a        │                0 │
│ t      │ b        │                0 │
│ t      │ c        │                0 │
│ c      │ default  │                0 │
│ t      │ default  │                8 │
│ t100   │ c        │                0 │
│ test   │ c        │                0 │
│ ccc    │ d        │                0 │
└──────────────────────────────────────┘
8 rows read in 0.134 sec. Processed 52 rows, 10.40 KiB (388.09 rows/s, 77.60 KiB/s)

root@localhost:8000/default>  select name,database,num_rows from system.tables where num_rows>=0;

┌──────────────────────────────────────┐
│  name  │ database │     num_rows     │
│ String │  String  │ Nullable(UInt64) │
├────────┼──────────┼──────────────────┤
│ t      │ a        │                0 │
│ t      │ b        │                0 │
│ t      │ c        │                0 │
│ t100   │ c        │                0 │
│ test   │ c        │                0 │
│ ccc    │ d        │                0 │
│ c      │ default  │                0 │
│ t      │ default  │                8 │
└──────────────────────────────────────┘
8 rows read in 0.183 sec. Processed 52 rows, 10.40 KiB (284.74 rows/s, 56.93 KiB/s)

root@localhost:8000/default> select version();

SELECT
  version()

-[ RECORD 1 ]-----------------------------------
version(): Databend Query v1.2.574-nightly-b643275639(rust-1.81.0-nightly-2024-08-02T21:24:00.176719947Z)

1 row read in 0.057 sec. Processed 1 row, 1B (17.52 rows/s, 17B/s)
```

Reply:
table = 'a' or database ='b' or num_rows = 80

Reply:
> table = 'a' or database ='b' or num_rows = 80

```sql
select name,database,num_rows from system.tables where num_rows=8 ;
+------+----------+----------+
| name | database | num_rows |
+------+----------+----------+
| t    | default  |        8 |
| v1   | default  |        8 |
+------+----------+----------+
2 rows in set (0.172 sec)

select name,database,num_rows from system.tables where name='t' 
    -> ;
+------+----------+----------+
| name | database | num_rows |
+------+----------+----------+
| t    | a        |        0 |
| t    | b        |        0 |
| t    | c        |        0 |
| t    | default  |        8 |
+------+----------+----------+
4 rows in set (0.196 sec)

select name,database,num_rows from system.tables where database='default' ;
+------+----------+----------+
| name | database | num_rows |
+------+----------+----------+
| c    | default  |        0 |
| t    | default  |        8 |
| v1   | default  |        8 |
+------+----------+----------+
3 rows in set (0.109 sec)

Union set:

select name,database,num_rows from system.tables where num_rows=8 or name='t' or database='default';
+------+----------+----------+
| name | database | num_rows |
+------+----------+----------+
| t    | default  |        8 |
| v1   | default  |        8 |
| t    | a        |        0 |
| t    | b        |        0 |
| t    | c        |        0 |
| c    | default  |        0 |
+------+----------+----------+
6 rows in set (0.184 sec)
```